language/bytecode-verifier/src/code_unit_verifier.rs

```diff
@@ -5,50 +5,84 @@
+    file_format::{
+        CompiledModule, CompiledScript, FunctionDefinition, FunctionDefinitionIndex,
+        IdentifierIndex,
+    },
     IndexKind,
 };
 
 pub struct CodeUnitVerifier<'a> {
-    module: &'a CompiledModule,
+    resolver: BinaryIndexedView<'a>,
+    function_view: FunctionView<'a>,
+    name_def_map: HashMap<IdentifierIndex, FunctionDefinitionIndex>,
 }
 
 impl<'a> CodeUnitVerifier<'a> {
-    pub fn verify(module: &'a CompiledModule) -> VMResult<()> {
-        let verifier = Self { module };
-        for (idx, function_definition) in verifier.module.function_defs().iter().enumerate() {
-            verifier
-                .verify_function(function_definition)
+    pub fn verify_module(module: &'a CompiledModule) -> VMResult<()> {
+        for (idx, function_definition) in module.function_defs().iter().enumerate() {
+            Self::verify_function(function_definition, module)
                 .map_err(|err| append_err_info(err, IndexKind::FunctionDefinition, idx))?
         }
         Ok(())
     }
 
-    fn verify_function(&self, function_definition: &FunctionDefinition) -> VMResult<()> {
+    pub fn verify_script(script: &'a CompiledScript) -> VMResult<()> {
+        // create `FunctionView` and `BinaryIndexedView`
+        let function_view = FunctionView::script(script);
+        let resolver = BinaryIndexedView::Script(script);
+        //verify
+        let code_unit_verifier = CodeUnitVerifier {
+            resolver,
+            function_view,
+            name_def_map: HashMap::new(),
+        };
+        code_unit_verifier.verify_common()
+    }
+
+    fn verify_function(
+        function_definition: &'a FunctionDefinition,
+        module: &'a CompiledModule,
+    ) -> VMResult<()> {
+        // nothing to verify for native function
         let code = match &function_definition.code {
-            Some(code) => &code.code,
+            Some(code) => code,
             None => return Ok(()),
         };
-
-        control_flow::verify(self.module, function_definition)?;
-        self.verify_function_inner(function_definition, &VMControlFlowGraph::new(code))
+        // create `FunctionView` and `BinaryIndexedView`
+        let function_handle = module.function_handle_at(function_definition.function);
+        let function_view = FunctionView::function(module, code, function_handle);
+        let resolver = BinaryIndexedView::Module(module);
+        let mut name_def_map = HashMap::new();
+        for (idx, func_def) in module.function_defs().iter().enumerate() {
+            let fh = module.function_handle_at(func_def.function);
+            name_def_map.insert(fh.name, FunctionDefinitionIndex(idx as u16));
+        }
+        // verify
+        let code_unit_verifier = CodeUnitVerifier {
+            resolver,
+            function_view,
+            name_def_map,
+        };
+        code_unit_verifier.verify_common()?;
+        AcquiresVerifier::verify(module, function_definition)
```

Comment:
While we are refactoring, the `AcquiresVerifier` should really be moved out into `verifier.rs` I was a noob when I added it, ands its definitely in the wrong spot

Reply:
I can do that

Reply:
ok I'll do this in another PR because I want to check better if combine with something else